lib/fast_jsonapi/object_serializer.rb

```diff
@@ -61,8 +61,9 @@ def hash_for_collection
       data = []
       included = []
       @resource.each do |record|
-        data << self.class.record_hash(record)
-        included.concat self.class.get_included_records(record, @includes, @known_included_objects) if @includes.present?
+        serializer = self.class.new(record)
```

Comment:
how does initializing an object with every iteration affect performance on large collections?

Reply:
Sharing numbers in current dev vs this branch:

dev:

```
Serialize to JSON string 250 records
Serializer      Records    Time
AMS serializer  250        83.96 ms
Fast serializer 250        3.4 ms

Serialize to Ruby Hash 250 records
Serializer      Records    Time
AMS serializer  250        77.04 ms
Fast serializer 250        2.73 ms

Serialize to JSON string 1000 records
Serializer      Records    Time
AMS serializer  1000       303.68 ms
Fast serializer 1000       9.94 ms

Serialize to Ruby Hash 1000 records
Serializer      Records    Time
AMS serializer  1000       300.06 ms
Fast serializer 1000       12.39 ms
```

this branch:

```
Serialize to JSON string 250 records
Serializer      Records    Time
AMS serializer  250        78.39 ms
Fast serializer 250        3.71 ms

Serialize to Ruby Hash 250 records
Serializer      Records    Time
AMS serializer  250        75.98 ms
Fast serializer 250        2.74 ms

Serialize to JSON string 1000 records
Serializer      Records    Time
AMS serializer  1000       307.92 ms
Fast serializer 1000       10.63 ms

Serialize to Ruby Hash 1000 records
Serializer      Records    Time
AMS serializer  1000       279.43 ms
Fast serializer 1000       8.98 ms
```

Reply:
More benchmarks (with includes and meta):

dev:

```
Serialize to JSON string 250 with includes and meta
Serializer      Records    Time
AMS serializer  250        112.38 ms
Fast serializer 250        4.62 ms

Serialize to Ruby Hash 250 with includes and meta
Serializer      Records    Time
AMS serializer  250        101.24 ms
Fast serializer 250        4.35 ms

Serialize to JSON string 1000 with includes and meta
Serializer      Records    Time
AMS serializer  1000       422.2 ms
Fast serializer 1000       20.34 ms

Serialize to Ruby Hash 1000 with includes and meta
Serializer      Records    Time
AMS serializer  1000       394.01 ms
Fast serializer 1000       18.07 ms
```

this branch:

```
Serialize to JSON string 250 with includes and meta
Serializer      Records    Time
AMS serializer  250        104.63 ms
Fast serializer 250        4.96 ms

Serialize to Ruby Hash 250 with includes and meta
Serializer      Records    Time
AMS serializer  250        101.09 ms
Fast serializer 250        4.3 ms

Serialize to JSON string 1000 with includes and meta
Serializer      Records    Time
AMS serializer  1000       401.42 ms
Fast serializer 1000       21.19 ms

Serialize to Ruby Hash 1000 with includes and meta
Serializer      Records    Time
AMS serializer  1000       389.07 ms
Fast serializer 1000       21.17 ms
```

Reply:
There is some cost, and incidentally, the performance cost using new instances on collections is something we actively debated about during development. On my machine the results are slightly different, and while the speed is impacted, it's still "performant" IMHO.

I do like the idea of @christophersansone's approach though. And I'm wondering if we can combine that approach with the instance method definition. The class could inspect for instance methods and convert them to procs using ruby goodness. Then, serialization would not require new instances and we could maintain the instance method syntax.

This is something I have not explored, and it may add more complexity than it warrants. It may also add some level confusion for developers on this project since an instance method would not be acting as an instance method... 

For giggles, here is a comparison in branches bumped to 10,000 records. I don't know why you would serialize 10,000 records following the JSON API spec, so it's a pretty contrived example, but it does illustrate the cost a little better.

dev:
```
Serialize to JSON string 10000 with includes and meta
Serializer      Records    Time
AMS serializer  10000      4791.46 ms
Fast serializer 10000      234.48 ms

Serialize to Ruby Hash 10000 with includes and meta
Serializer      Records    Time
AMS serializer  10000      4484.9 ms
Fast serializer 10000      193.3 ms
```

this branch:
```
Serialize to JSON string 10000 with includes and meta
Serializer      Records    Time
AMS serializer  10000      4741.36 ms
Fast serializer 10000      360.68 ms

Serialize to Ruby Hash 10000 with includes and meta
Serializer      Records    Time
AMS serializer  10000      4529.02 ms
Fast serializer 10000      214.24 ms
```

Reply:
I also wonder if it's better to take a performance hit by merging this in, and then iterating to a more performant approach.

Reply:
> It may also add some level confusion for developers on this project since an instance method would not be acting as an instance method

+1 on that.  We would definitely have to take into account attribute methods that call other methods (e.g. if you have a complex method to calculate an attribute that calls other methods).  A developer would expect that to just work, so just cherry-picking certain instance methods could be trouble.  And using instance variables would not behave as expected, either.  The serializer could potentially have some assertions for the developer to help avoid some common pitfalls.  But if we can avoid all of this by just having a block with the attribute definition, that works.

A couple different options for those who would really want a separate method (though they would have to be class methods):

```ruby
attribute :title_with_year { |r| MovieSerializer.title_with_year(r) }

def self.title_with_year(record)
  ...
end
```

or

```ruby
attribute :title_with_year, class_method: :title_with_year

def self.title_with_year(record)
  ...
end
```

Or we just have a generator function (or recommend that they do it themselves).  Or we just provide the capability and offer some suggestions in the readme...

```ruby
def self.class_attribute(attr)
  attribute(attr) { |record| self.class.send(attr, record) }
end

class_attribute :title_with_year
```

Lots of options to avoid serializer instantiation.  My opinion would be to start with the simplest, most performant solution and then iterate on it based on feedback or limitations.

Reply:
@christophersansone @grossadamm I like the simplicity of the Proc approach but in my opinion, there are many cases where it's convenient to break the calculation of attributes into multiple methods and I'm not sure if we can achieve that using Procs. e.g:

```ruby
class FeedItemSerializer
  # ignoring other attributes
  attributes :cooked_count, :likes_count, :liked_photo_comments_count

  # ignoring other associations
  has_many :liked_photo_comments

  def liked_photo_comments
    return [] unless triggered_by_like?
    photo_comments_liked_by_author.recent.limit(4)
  end

  def liked_photo_comments_count
    return unless triggered_by_like?
    photo_comments_liked_by_author.count
  end

  def cooked_count
    approved_photo_comments.count
  end

  def likes_count
    recipe_likes.count
  end

  private

    def triggered_by_like?
      # ...
    end

    def photo_comments_liked_by_author
      # ...
    end

    def recipe_likes
      # ...
    end

    def approved_photo_comments
      # ...
    end
end
```

Reply:
Another case we rely on method attributes and goodies of OOP is in our fragment caching (used to cache attributes in serializers where isn't desirable to cache entire objects)

```ruby
# Using Rails.cache for the example, the logic of our Caching System is a bit more complex.
class ApplicationSerializer < ActiveModel::Serializer
  def self.expire_cache_fragment!(name)
    Rails.cache.delete("ams_fragment_cache_#{name}")
  end

  protected
  def cache_fragment(name)
    Rails.cache.fetch("ams_fragment_cache_#{name}") do
      yield
    end
  end
end
```
And then we can use the `cache_fragment` method in all the subclasses of ApplicationSerializer, e.g:

```ruby
class MovieSerializer < ApplicationSerializer
  attributes :background_actors

  def background_actors
    cache_fragment("background_actors") do
      # Heavy computation to serialize all background actors ...
    end
  end
end
```

Reply:
@guilleiguaran @grossadamm Agreed that complex calculations could be cumbersome with the block syntax, and the AMS syntax is quite nice for doing this.  But is it worth the performance hit for every single serialization, regardless of whether this extra syntactic sugar is needed?  My opinion is that we should not force some performance penalty at all times, but potentially allow the developer to opt into it as needed.

One solution is to simply use class methods, passing in the record each time.  This example assumes attribute block syntax and nothing else, but we can design something better if the concept is a good one:

```ruby
attribute :title_with_year { |r| MovieSerializer.title_with_year(r) }

def self.title_with_year(record)
  ...
end
```

If that is not sufficient, then consider having a separate decorator class that can be instantiated.

```ruby
attribute :title_with_year { |r| MovieDecorator.new(r).title_with_year }
```

We could also implement the ability to define a decorator in the serializer, which would instantiate an instance of the decorator for each record, e.g.:

```ruby
class MovieSerializer
  include FastJsonapi::ObjectSerializer
  decorator MovieDecorator
  attribute :title_with_year
end

decorator MovieDecorator
  include FastJsonapi::ObjectDecorator
  def title_with_year
    ...
  end
end
```

The decorator would basically define an `initialize` with the record, and it would implement `method_missing` to send all undefined methods to the record.

Lots of options here, just trying to come up with some ideas that work but don't impose any performance loss or unnecessary memory consumption by default.